sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/validation/QueryResults.kt

```diff
@@ -42,19 +45,38 @@ data class QueryResults private constructor(
      */
     internal val tables: Map<String, QueryTable>
 ) {
-  internal val requiresType = columns.size + tables.size > 1
+  private val interfaceClassName = "${queryName.capitalize()}Model"
 
-  internal fun generateTypeSpec(): TypeSpec {
-    val typeSpec = TypeSpec.interfaceBuilder("${queryName.capitalize()}Model")
-        .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+  internal val requiresType = columns.size + tables.size > 1
 
-    columns.map { it.value.index to it.value.interfaceMethod(it.key) }
-        .plus(tables.map { it.value.index to it.value.interfaceMethod(it.key) })
-        .sortedBy { it.first }
-        .forEach { typeSpec.addMethod(it.second) }
-
-    return typeSpec.build()
-  }
+  private fun <T> sortedResultsMap(
+      columnsMap: (String, IndexedValue) -> T,
+      tablesMap: (String, QueryTable) -> T
+  ) = columns.map { it.value.index to columnsMap(it.key, it.value) }
+      .plus(tables.map { it.value.index to tablesMap(it.key, it.value) })
+      .sortedBy { it.first }
+      .map { it.second }
+
+  internal fun generateInterface() = TypeSpec.interfaceBuilder(interfaceClassName)
+      .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+      .addMethods(sortedResultsMap(
+          { columnName, value -> value.interfaceMethod(columnName) },
+          { tableName, table -> table.interfaceMethod(tableName) }
+      ))
+      .build()
+
+  internal fun generateCreator() = TypeSpec.interfaceBuilder("${queryName.capitalize()}Creator")
+      .addTypeVariable(TypeVariableName.get("T", ClassName.bestGuess(interfaceClassName)))
```

Comment:
This `ClassName.bestGuess` is a bit sketchy and a bit of a hack. We should figure out a way to avoid this in the future and always use qualified names. Not super important since we know they're always in the same file but over time I've found that my uses of `bestGuess` were always a sign of a modeling problem.